207 pinpoint server implementation

robertvo824 opened the pull request:
## Overview

[analytics](https://github.com/DavidWells/analytics) and [@analytics/aws-pinpoint](https://github.com/DavidWells/analytics/tree/master/packages/analytics-plugin-aws-pinpoint) is used in [telemetry.js](https://devtopia.esri.com/WebGIS/arcgis-telemetry.js) to send data to Pinpoint and eventually gets stored in Redshift. The current implementation of the `analytics pinpoint` package works for sending data from browser, but currently does not have `Node.js` implementation. This forked repo helps resolve that so that we can use `telemetry.js` server-side for aws.

## Technical

- In the original repo, there is a [browser implementation](https://github.com/DavidWells/analytics/blob/master/packages/analytics-plugin-aws-pinpoint/src/browser.js) for Pinpoint but not for [Node](https://github.com/DavidWells/analytics/blob/master/packages/analytics-plugin-aws-pinpoint/src/node.js)
- This PR took the code from the `browser.js` and pasted into the `node.js` and resolved issues
- Duplicated logic was pulled into shared modules
- Used `inBrowser` module to check if the package is running client vs server-side

## Description

- Working state of `@analytics/aws-pinpoint` to send custom events from `Node.js` to `AWS Pinpoint`.
- Still refactoring, moving things around, and cleaning unnecessary logic

robertvo824 commented:
@rgwozdz 

- Moved helpers into separate directory
- Created one Pinpoint `index.js` file (without having a `browser.js` and `node.js`)
    - Created one Pinpoint `call-aws.js` file that handles the creation of aws-client using `aws4fetch` or `aws-sdk` depending on `inBrowser`

rgwozdz commented on packages/analytics-plugin-aws-pinpoint/src/node.js:
```diff
@@ -1,9 +1,187 @@
+
+      if (userId) {
+        userInfo.UserId = userId
+      }
```
I notice in browser.js at this location, there is a block:
```
      if (traits.email) {
        ;(endpoint.Address = traits.email),
          (endpoint.ChannelType = CHANNEL_TYPES.EMAIL)
      }
```
why was that removed from the node.js version?

robertvo824 replied:
Sorry, forgot to reply to this comment. That was from my initial attempt at getting it to work in Node. I didn't want to focus on email at the time. I will go ahead and add it back in when I pull out some shared code.

rgwozdz commented on packages/analytics-plugin-aws-pinpoint/src/node.js:
```diff
@@ -1,9 +1,187 @@
+  throw new Error('Pinpoint not loaded')
+}
+
+function formatEventData(obj) {
```
I think this could be shared code; appears in browser.js too

robertvo824 replied:
Good call. This was the first file I touched so I didn't event think about sharing code early on.

rgwozdz commented on packages/analytics-plugin-aws-pinpoint/src/node.js:
```diff
@@ -1,9 +1,187 @@
   }
 }
+
+function loadError() {
```
I think this could be shared code, appears in browser.js too.

rgwozdz commented on packages/analytics-plugin-aws-pinpoint/src/node.js:
```diff
@@ -1,9 +1,187 @@
+      /* Initialize session info */
+      const initSessionData = getSession()
+      logger('initSessionData', initSessionData)
+      /* If anonId has changed, refresh session details */
```
Add line of whitespace

rgwozdz commented on packages/analytics-plugin-aws-pinpoint/src/node.js:
```diff
@@ -1,9 +1,187 @@
+      ...config,
+      ...pluginConfig,
+    },
+    bootstrap: (pluginApi) => {
```
I think this could be shared code, appears in browser.js too.

rgwozdz reviewed (changes requested):
This is excellent work Bob.  Great job extracting shared code to helpers!  There may be a few more opportunities for that, see my comments.

Are there no tests for the original browser code?  Seems like we'd want to make sure the new node pieces are under test.

robertvo824 commented:
> Are there no tests for the original browser code? Seems like we'd want to make sure the new node pieces are under test.

That is correct. It appears that he lacks tests for all of the individual plugins (browser and node), but contains tests for `analytics-core` and his `utils` which I believe are supposed to cover any and all plugins that are added. For example, this is the file for server-side test in `analytics-core` https://github.com/DavidWells/analytics/blob/master/packages/analytics-core/tests/api-serverside.test.js.

rgwozdz commented:
On tests...ok, that's pretty wild.  To me that means you can submit your fork back to the main repo and see what the response is.  I do think we need a follow up task to put everything under test.

robertvo824 commented:
Made requested changes:

- Moved shared logic for loadError, formatEventData, and bootstrap into separate modules
- Added line of whitespace
- Fixed `session_start is undefined` error in `create-pinpoint-sender.js`

robertvo824 commented:
Closing this out and made a new PR on the original repo.

https://github.com/DavidWells/analytics/pull/217
